[FEATURE] Compresser les exports CPF (PIX-5399)

aceol opened the pull request:
## :unicorn: Problème
Les exports CPF limités à 256mo, sont uploadés sur un bucket avant d'être transmis au CPF. Le stockage et les transferts sont couteux.

## :robot: Solution
Compresser les exports avant d'être stockés

## :rainbow: Remarques
- La compression se fait nativement avec node via les streams

## :100: Pour tester
Générer un export. S'assurer qu'il est compressé sur le bucket

- s'assurer que les lignes ne sont pas flaggées comme traitées:
` update "certification-courses" set "cpfFilename"= null` -- 50014 lignes
- Verifier le nombre de lignes par fichiers: CPF_PLANNER_JOB_CHUNK_SIZE // 50000 pour avoir 1 fichier, 20000 pour avoir 3 fichiers
- Mettre à jour le cron: CPF_PLANNER_JOB_CRON = MM HH * * *
- Verifier la generation des fichiers dans les logs
- Recuperer les fichiers sur le[ bucket ovh ](https://www.ovh.com/manager/#/public-cloud/pci/projects/27c5a6d3d35841a5914c7fb9a8e96345/storages/objects/63476c344c574e775a69316b5a58597552314a42)
- Compter pour chaque fichier les lignes générées, qui doivent correspondre aux nombre attendus de certif / CPF_PLANNER_JOB_CHUNK_SIZE
`grep -o "<cpf:passageCertification>" pix-cpf-export-20221011-092517332.xml | wc -l`

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4982.review.pix.fr
- App (.org): https://app-pr4982.review.pix.org
- Orga: https://orga-pr4982.review.pix.fr
- Certif: https://certif-pr4982.review.pix.fr
- Admin: https://admin-pr4982.review.pix.fr
- API: https://api-pr4982.review.pix.fr/api/

 Please check it out!

P-Jeremy commented on api/tests/integration/infrastructure/jobs/cpf-export/handlers/create-and-upload_test.js:
```diff
@@ -0,0 +1,97 @@
+      //const fs = require('fs');
+      //const writableStreamToFile = fs.createWriteStream(filename);
+      //writableStream.pipe(writableStreamToFile);
+      const unzip = createUnzip();
```
Commentaires :)

aceol replied:
Je voulais les laisser pour qu'on puisse generer le fichier pour tester. 
Mais on peut les supprimer

P-Jeremy replied:
Ah ok, tu avais fait un truc pas mal sur l'attestation pdf, une sous méthode qui fait le boulot, et qui prends un boolean en paramètre.

api/tests/integration/infrastructure/utils/pdf/certification-attestation-pdf_test.js

![image](https://user-images.githubusercontent.com/37305474/192478533-37efb0ae-74df-4c29-93c8-511af4f3730f.png)

aceol replied:
Yes mais la, c'est dans les tests, je vais plutot le supprimer et mettre en commentaire de commit

P-Jeremy commented on api/tests/integration/infrastructure/jobs/cpf-export/handlers/create-and-upload_test.js:
```diff
@@ -0,0 +1,97 @@
+      //writableStream.pipe(writableStreamToFile);
+      const unzip = createUnzip();
+      const unziped = writableStream.pipe(unzip); //.pipe(process.stdout);
+      const readableFile = await _streamToString(unziped);
```
Idem

aceol commented:
~~La PR est bloqué en attendant plus de tests fonc~~

aceol committed:
:recycle: api: domainBuilder default date is string

aceol committed:
:recycle: api: Control stream drain

stream.write return false if it needs to drain before next write

aceol committed:
:recycle: Rename writableStream wich is a readableStream

mariannebost commented on api/lib/infrastructure/jobs/cpf-export/handlers/create-and-upload.js:
```diff
@@ -1,27 +1,54 @@
-    limit,
-    offset,
-  });
+  try {
```
Vérifier le try

mariannebost commented on api/lib/infrastructure/jobs/cpf-export/handlers/create-and-upload.js:
```diff
@@ -1,27 +1,54 @@
+    const filename = `pix-cpf-export-${now}.xml.gz`;
+    await cpfExternalStorage.upload({
+      filename,
+      writableStream: gzipStream,
```
readable et pas writable

mariannebost commented on api/tests/integration/infrastructure/jobs/cpf-export/handlers/create-and-upload_test.js:
```diff
@@ -0,0 +1,90 @@
+      cpfExternalStorage,
+    });
+
+    expect(cpfCertificationResultRepository.markCertificationCoursesAsExported).to.have.been.calledWith({
```
// then

aceol committed:
:sparkles: api: GZIP cpf stream

aceol committed:
:recycle: api: Replace moment by dayjs

aceol committed:
:sparkles: api: Manage timezone

Test where failing when timezone was different
for instance: `TZ=Asia/Kolkata npm t`
